### ISSUE TITLE
Type inference issue with match expression in generic functions

### ISSUE THREAD

**Issue by myuanz** [microsoft/pyright#5777]
**Describe the bug**
Pyright is unable to correctly handle type inference within generic functions that utilize the match expression, leading to type errors when returning different types of objects. In the provided example code, the match expression is used to match different classes and return corresponding instances.

**Code or Screenshots**

```python
from typing import TypeVar
from dataclasses import dataclass

@dataclass
class Test1:
    a: int

@dataclass
class Test2:
    b: int

T = TypeVar('T', Test1, Test2)

def f(t: T) -> T:
    match t:
        case Test1(a):
            return Test1(a+1) # Expression of type "Test1" cannot be assigned to return type "T@f"
        case Test2(b):
            return Test2(b-1) # Expression of type "Test2" cannot be assigned to return type "T@f"
        case _ as unreachable:
            raise TypeError()

def g(t: T) -> T:
    if isinstance(t, Test1):
        return Test1(t.a + 1) # OK
    elif isinstance(t, Test2):
        return Test2(t.b - 1) # OK
    else:
        raise TypeError()

@dataclass
class Test3:
    c: Test2

U = TypeVar('U', int, Test2, Test3)
def h(u: U):
    print(u)
    if isinstance(u, int):
        return u - 1 # OK
    elif isinstance(u, Test2):
        return Test2(b=h(u.b)) # error: Argument of type "int* | Test2 | Test3" cannot be assigned to parameter "b" of type "int" in function "__init__"
    elif isinstance(u, Test3):
        return Test3(c=h(u.c)) # OK
    else:
        raise ValueError()

```

The problem of function `f` also exists in mypy(https://github.com/python/mypy/issues/15930), but the problem of function `h` does not exist in mypy. I think the functions `f` and `h` may be two kinds of errors, but for convenience, they are put together. 

**VS Code extension or command-line**
Triggered in the following two environments:

1. Python 3.10.12 + pyright 1.1.323 cmd
2. Python 3.11.3 + VS Code 1.81.1 + Pylance v2023.8.31

**Comment by erictraut** [comment]
You're using a constrained (sometimes called "restricted") TypeVar here. This construct is unique to Python (you won't find it in any other language, to my knowledge), and it's a difficult feature to support. I wish I could go back in time and prevent it from being added to the Python type system, but it's unfortunately too late to do so.

I generally recommend avoiding the use of constrained TypeVars if you can help it. There are typically better approaches, such as the use of unions and overloads. For example, if I were writing the code above, I would define `f` as follows:

```python
@overload
def f(t: Test1) -> Test1: ...

@overload
def f(t: Test2) -> Test2: ...

def f(t: Test1 | Test2) -> Test1 | Test2:
    match t:
        case Test1(a):
            return Test1(a + 1)
        case Test2(b):
            return Test2(b - 1)
        case _ as unreachable:
            raise TypeError()
```

There are limits to what pyright can do in supporting constrained TypeVars. Refer to [this documentation](https://microsoft.github.io/pyright/#/mypy-comparison?id=constrained-type-variables) for more insights into the associated challenges.

Mypy's approach to handling constrained TypeVars is different from pyright's, and it has its own (different) set of problems and false positives. I recently reviewed all of the open issues in the mypy issue tracker, and I came across at least a dozen open issues relating to constrained TyepVars.

I'll spend some time to see if I can make the code sample above type check in pyright, but I want to set expectations that the answer might be "close as won't fix".

**Comment by erictraut** [comment]
For the last example (function `h`), the error occurs because pyright isn't able to infer the return type of the function. That's not uncommon for recursive functions, and it can be easily solved by supplying a return type annotation. Mypy doesn't ever attempt to infer function return types, so it assumes the return type is `Any` in this case. If you provide an explicit return type of `-> U`, that problem goes away. So I don't consider that a bug.

**Comment by erictraut** [comment]
I've added extra logic to handle the case where the return type is a constrained TypeVar and the return statement is guarded by a `case` statements using a class pattern. This will be included in the next release of pyright.

**Comment by erictraut** [comment]
This is included in pyright 1.1.324, which I just published. It will also be included in a future release of pylance.

**Comment by myuanz** [comment]
Thank you for your prompt response. I didn't realize that `TypeVar` could be this tricky; I had assumed that `TypeVar('T', Test1, Test2)` was somewhat analogous to TypeScript's `type T = Test1 | Test2`.

I've attempted to rewrite it using the `@overload` approach locally, and so far, it seems to be working without any issues, albeit with a bit more `type`.

Best regards
